code/modules/mining/machine_silo.dm

```diff
@@ -77,93 +77,74 @@ GLOBAL_LIST_EMPTY(silo_access_logs)
-			if (sheets >= 20)
-				ui += "<a href='?src=[REF(src)];ejectsheet=[ref];eject_amt=20'>20x</a>"
-			else
-				ui += "<span class='linkOff'>20x</span>"
-			ui += "<b>[mat.name]</b>: [sheets] sheets<br>"
-			any = TRUE
-	if(!any)
-		ui += "Nothing!"
-
-	ui += "</div><div class='statusDisplay'><h2>Connected Machines:</h2>"
-	for(var/datum/component/remote_materials/mats as anything in ore_connected_machines)
-		var/atom/parent = mats.parent
-		ui += "<a href='?src=[REF(src)];remove=[REF(mats)]'>Remove</a>"
-		ui += "<a href='?src=[REF(src)];hold=[REF(mats)]'>[holds[mats] ? "Allow" : "Hold"]</a>"
-		ui += " <b>[parent.name]</b> in [get_area_name(parent, TRUE)]<br>"
-	if(!ore_connected_machines.len)
-		ui += "Nothing!"
-
-	ui += "</div><div class='statusDisplay'><h2>Access Logs:</h2>"
-	var/list/logs = GLOB.silo_access_logs[REF(src)]
-	var/len = LAZYLEN(logs)
-	var/num_pages = 1 + round((len - 1) / 30)
-	var/page = clamp(log_page, 1, num_pages)
-	if(num_pages > 1)
-		for(var/i in 1 to num_pages)
-			if(i == page)
-				ui += "<span class='linkOff'>[i]</span>"
-			else
-				ui += "<a href='?src=[REF(src)];page=[i]'>[i]</a>"
-
-	ui += "<ol>"
-	any = FALSE
-	for(var/i in (page - 1) * 30 + 1 to min(page * 30, len))
-		var/datum/ore_silo_log/entry = logs[i]
-		ui += "<li value=[len + 1 - i]>[entry.formatted]</li>"
-		any = TRUE
-	if (!any)
-		ui += "<li>Nothing!</li>"
-
-	ui += "</ol></div>"
-	return ui.Join()
-
-/obj/machinery/ore_silo/Topic(href, href_list)
-	if(..())
+/obj/machinery/ore_silo/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/spritesheet/sheetmaterials)
+	)
+
+/obj/machinery/ore_silo/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "OreSilo")
+		ui.open()
+
+/obj/machinery/ore_silo/ui_static_data(mob/user)
+	return materials.ui_static_data()
+
+/obj/machinery/ore_silo/ui_data(mob/user)
+	var/list/data = list(
+		"materials" =  materials.ui_data()
+	)
+
+	var/list/connected_data
+	for(var/datum/component/remote_materials/remote as anything in ore_connected_machines)
+		var/atom/parent = remote.parent
+		var/icon/parent_icon = icon(initial(parent.icon), initial(parent.icon_state), frame = 1)
+		var/list/remote_data = list(
+			"ref" = REF(remote),
+			"icon" = icon2base64(parent_icon),
+			"name" = parent.name,
+			"onHold" = !!holds[remote],
+			"location" = get_area_name(parent, TRUE)
+		)
+		LAZYADD(connected_data, list(remote_data))
+	LAZYSET(data, "machines", connected_data)
+
+	var/list/logs_data
+	for(var/datum/ore_silo_log/entry as anything in GLOB.silo_access_logs[REF(src)])
+		var/list/log_data = list(
+			"rawMaterials" = entry.get_raw_materials(""),
+			"machineName" = entry.machine_name,
+			"areaName" = entry.area_name,
+			"action" = entry.action,
+			"amount" = entry.amount,
+			"time" = entry.timestamp,
+			"noun" = entry.noun
+		)
+		LAZYADD(logs_data, list(log_data))
+	LAZYSET(data, "logs", logs_data)
```

Comment:
Why bother with lazy lists? Just makes this code more complicated.

Reply:
it's free memory